### ISSUE TITLE
Boost variant adaptor is broken: which() may return 0

### ISSUE THREAD

**Issue by Rupsbant** [solodon4/Mach7#73]
The current implementation of vtblmap4.hpp initialises and assumes vtbl indexes are not 0.
The boost adaptor may return a vtbl index of 0 when storing the first type of the variant.

1. https://github.com/solodon4/Mach7/blob/40f062d21d721e251ba15c5e8e99068129a6bacf/code/mach7/vtblmap4.hpp#L253
2. https://github.com/solodon4/Mach7/blob/40f062d21d721e251ba15c5e8e99068129a6bacf/code/mach7/vtblmap4.hpp#L264
3. https://github.com/solodon4/Mach7/blob/40f062d21d721e251ba15c5e8e99068129a6bacf/code/mach7/vtblmap4.hpp#L270
4. https://github.com/solodon4/Mach7/blob/40f062d21d721e251ba15c5e8e99068129a6bacf/code/mach7/adapters/boost/adapt_boost_variant.hpp#L62

This results in an error when first matching against an index 0 and then having a collision. Which will reuse the cache-entry with a dirty switch_info.

I see three fixes:

1. Boost adaptor: change code to p->which() + 1;
2. Initialize cache entries with -1 value instead of 0. (or use some value outside the 48-bit pointer range)
3. Add a dirty flag to the cache entry to test isOccupied correctly.

I think option 2 is the best if possible. Option 1 keeps the bug around for other adaptors. Option 3 has a clear overhead.

Minimal example and pull request follow.

**Comment by akrzemi1** [comment]
However, this will not fix all the problems with `boost::variant` support. `boost::variant` does not guarantee that its contained value is always at the same offset relative to the address of variant; and such guarantee is required by `Mach7`.

**Comment by solodon4** [comment]
Thanks for reporting the issue. @Rupsbant. I'd prefer the solution 1 as it is contained within adapter and thus has the least chances of messing up the case for polymorphic classes, for which the timings in the paper were reported. I'm pretty sure there will be a number of places I assumed non-nullness of v-table pointers and used that fact.

It is also true that currently Mach7 assumes fixed offset from the object to sub-object, as @akrzemi1 points out, however this is not a hard requirement. I've been planning to provide a customization point that would prevent memorizing offsets in order to be able to handle boost::any. This is currently tracked by issue #17

**Comment by Rupsbant** [comment]
Minimal example: 
```
#include <mach7/adapters/boost/adapt_boost_variant.hpp>
#include <mach7/patterns/constructor.hpp> 
#include <mach7/patterns/primitive.hpp>   

template<int N> struct A {
  int val;
  A(int a) : val(a) {};
};
struct CC {
  boost::variant<A<0>, A<1>, A<2>, A<3>, A<4>, A<5>, A<6>, A<7>, A<8>> variant;
  template<typename T> CC(T t) : variant(t) {};
};
namespace mch {
  template <int N> struct bindings<A<N>> {
    Members(A<N>::val);
  };
  template <> struct bindings<CC> {
    Members(CC::variant);
  };
}
void do_some(CC c) {
  using namespace mch;
  var<int> x;
  Match(c.variant) {
    Case(C<A<0>>(x)) {
      std::cout << 0 << " " << x << std::endl;
      break;
    }
    Case(C<A<8>>(x)) {
      std::cout << 8 << " " << x << std::endl;
      break;
    }
    Otherwise() {
      std::cout << "Other" <<std::endl;
      break;
    }
  }
  EndMatch;
}
template<int N> void test() {
  do_some(CC(A<N>(5)););
}
int main() {
  test<0>();
  test<8>();
}
```
Current output: 
```
0 5
0 5
```
Currently the layout of A<0> and A<8> are the same so there is no segfault.
Expected output:
```
0 5
8 5
```

**Comment by Rupsbant** [comment]
Boost::variant currently stores the data with a union type inside the object not via indirection. The offsets are thus fixed like assumed in issue #17. 

I don't think boost::variant would ever use a heap based implementation, it's only a little bit more type safe than boost::any with the almost the same runtime. Now they have clearly different use-cases.

**Comment by solodon4** [comment]
I've added your above test as unit test and merged in your pull request. Thank you for fixing this!
